rascal-lsp/src/main/rascal/demo/lang/pico/LanguageServer.rsc

```diff
@@ -81,31 +104,73 @@ Summary picoSummarizer(loc l, start[Program] input, PicoSummarizerMode mode) {
     return s;
 }
 
+@synopsis{Finds a declaration that the cursor is on and proposes to remove it.}
+list[CodeAction] picoActions([*_, IdType x, *_, start[Program] program]) 
+    = [action(command=removeDecl(program, x, title="remove <x>"))];
+
+default list[CodeAction] picoActions(list[value] _) = [];
+
 set[loc] lookupDef(loc _, start[Program] input, Tree cursor) =
     { d.src | /IdType d := input, cursor := d.id};
 
-
+@synsopsis{Defines three example commands that can be triggered by the user (from a code lens, from a diagnostic, or just from the cursor position)}
 data Command
-  = renameAtoB(start[Program] program);
+  = renameAtoB(start[Program] program)
+  | changeToFix(loc place, str replacement)
+  | removeDecl(start[Program] program, IdType toBeRemoved)
+  ;
 
-rel[loc,Command] picoLenses(start[Program] input) = {<input@\loc, renameAtoB(input, title="Rename variables a to b.")>};
+@synopsis{Adds an example lense to the entire program.}
+rel[loc,Command] picoLenses(start[Program] input) 
+    = {<input@\loc, renameAtoB(input, title="Rename variables a to b.")>};
 
 
+@synopsis{Generates inlay hints that explain the type of each variable usage.}
 list[InlayHint] picoHinter(start[Program] input) {
     typeLookup = ( "<name>" : "<tp>" | /(IdType)`<Id name> : <Type tp>` := input);
+
     return [
-        hint(name.src, " : <typeLookup["<name>"]>", \type(), atEnd = true) | /(Expression)`<Id name>` := input, "<name>" in typeLookup
+        hint(name.src, " : <typeLookup["<name>"]>", \type(), atEnd = true) 
+        | /(Expression)`<Id name>` := input
+        , "<name>" in typeLookup
     ];
 }
 
+@synopsis{Helper function to generate actual edit actions for the renameAtoB command}
 list[DocumentEdit] getAtoBEdits(start[Program] input)
    = [changed(input@\loc.top, [replace(id@\loc, "b") | /id:(Id) `a` := input])];
 
+@synopsis{Command handler for the renameAtoB command}
 value picoCommands(renameAtoB(start[Program] input)) {
     applyDocumentsEdits(getAtoBEdits(input));
     return ("result": true);
 }
 
+@synopsis{Command handler for the changeToFix command}
+value picoCommands(changeToFix(loc src, str newName)) {
+    applyDocumentsEdits([changed(src.top, [replace(src, newName)])]);
+    return ("result": true);
+}
+
+@synopsis{Command hamlder for the removeDecl command}
+value picoCommands(removeDecl(start[Program] program, IdType toBeRemoved)) {
+    applyDocumentsEdits([changed(program@\loc.top, [replace(toBeRemoved@\loc, "")])]);
+    return ("result": true);
+}
+
+@synopsis{The main function registers the Pico language with the IDE}
+@description{
+Register the Pico language and the contributions that supply the IDE with features.
+
+((registerLanguage)) is called twice here:
+1. first for fast and cheap contributions
+2. asynchronously for the full monty that loads slower
```

Comment:
not sure if it's more async than the other? just on a complelty different evaluator as to not impact the other?

Reply:
For the user it is important to know that either call does not block the other, so "asynchronously" points to the relation between the first and the second call.